File: host/react/src/underlay.ts
// The shadow underlay — where a floating surface's depth actually lives.
//
// A tile is a webview, and a webview is clipped to its own rect: a shadow drawn
// by the program inside it casts into nothing. So the aura is drawn *beneath*
// the tiles instead, by one full-window webview the rim creates first (bottom of
// the z-order) and never lets take a click. This module is that webview's whole
// page: it imports the shared token stylesheet, so styling stays here rather
// than crossing into Rust, and it exposes one global the rim calls with the
// only thing the rim knows — the tile leaves' rectangles, in window coordinates.
//
// It is not a program: no process, no scope, no engine. It is the canvas the
// window's depth is painted on.
import { styles } from './styles.ts'

/** One tile leaf's geometry, as the rim's `geometry::walk` hands it over. */
export type Rect = { x: number; y: number; width: number; height: number }

declare global {
  interface Window {
    /** Called by the rim on mount and on every layout change. */
    __openlight_underlay?: (rects: Rect[]) => void
    /** What the rim said before this module loaded (its init script buffers). */
    __openlight_rects?: Rect[]
  }
}

/** One aura per rect, positioned where the tile is — pure but for the DOM. */
export function render(host: HTMLElement, rects: Rect[]): void {
  host.replaceChildren(
    ...rects.map((rect) => {
      const aura = document.createElement('div')
      aura.dataset.ui = 'aura'
      aura.style.left = `${rect.x}px`
      aura.style.top = `${rect.y}px`
      aura.style.width = `${rect.width}px`
      aura.style.height = `${rect.height}px`
      return aura
    }),
  )
}

// Layout only — the radius and the aura itself are tokens in `styles.ts`.
const CSS = `
  /* Nothing here is ever a target: the underlay spans the whole window, and the
     surfaces it sits under must keep every click. */
  html, body { pointer-events: none }
  [data-ui="underlay"] { position: fixed; inset: 0 }
  /* An outer box-shadow never paints inside the box that casts it, so the aura
     is a halo around the tile and nothing else — the tile's own white card,
     edge to edge in its webview, covers the rest. */
  [data-ui="aura"] {
    position: absolute;
    border-radius: var(--ol-radius); box-shadow: var(--ol-shadow-aura);
  }
`

const sheet = document.createElement('style')
sheet.textContent = styles + CSS
document.head.appendChild(sheet)

const host = document.createElement('div')
host.dataset.ui = 'underlay'
document.body.appendChild(host)

const buffered = window.__openlight_rects ?? []
window.__openlight_underlay = (rects: Rect[]) => render(host, rects)
window.__openlight_underlay(buffered)

File: host/src/underlay.rs
//! The shadow underlay, rim side — the surface a tile's aura is cast on.
//!
//! A webview is clipped to its own rect, so a floating surface cannot draw the
//! shadow that proves it floats. The rim therefore creates one more webview
//! before any other: full-window, transparent, click-through, at the bottom of
//! the z-order. Its page is `react/src/underlay.ts`, served over `ol://` like
//! any module, and it draws a rounded aura at each tile leaf's geometry.
//!
//! Everything visual belongs to that page. The rim's whole vocabulary here is
//! *rectangles* — [`script`] is the only thing it ever says.

use crate::geometry::Rect;
use serde_json::{json, Value};

/// The id the underlay's webview is registered under. It is not a process: no
/// engine run ever produces this, so it can never collide with one.
pub const PROCESS: &str = "__underlay";

/// What the probe lane calls it — the underlay answers like any surface.
pub const PROGRAM: &str = "underlay";

/// The page, relative to the host project root (the base `body.executable`
/// resolves from, `boot::Booted::programs_root`).
pub const ENTRY: &str = "react/src/underlay.ts";

/// Installed before the page's own code (`with_initialization_script`), once
/// per document.
///
/// The layout is known before the page exists, and an `evaluate_script` sent
/// into a document that has not loaded is simply lost — so the first rects ride
/// in here, where the page is guaranteed to find them, and the global that will
/// carry every later one is defined buffering until the module claims it.
pub fn init_script(rects: &[Rect]) -> String {
    format!(
        "window.__openlight_rects = {};\n\
         window.__openlight_underlay = (rects) => {{ window.__openlight_rects = rects; }};",
        payload(rects)
    )
}

/// The one message the rim sends a loaded underlay: where the tiles are.
pub fn script(rects: &[Rect]) -> String {
    format!("window.__openlight_underlay({});", payload(rects))
}

fn payload(rects: &[Rect]) -> Value {
    Value::Array(
        rects
            .iter()
            .map(|r| json!({ "x": r.x, "y": r.y, "width": r.width, "height": r.height }))
            .collect(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_rim_says_rectangles_and_nothing_else() {
        let js = script(&[
            Rect { x: 240.0, y: 14.0, width: 1000.0, height: 800.0 },
            Rect { x: 10.5, y: 0.0, width: 1.0, height: 2.0 },
        ]);
        let payload = js
            .strip_prefix("window.__openlight_underlay(")
            .and_then(|rest| rest.strip_suffix(");"))
            .expect("one call, one argument");
        let sent: Value = serde_json::from_str(payload).expect("the argument is data");
        assert_eq!(
            sent,
            json!([
                { "x": 240.0, "y": 14.0, "width": 1000.0, "height": 800.0 },
                { "x": 10.5, "y": 0.0, "width": 1.0, "height": 2.0 },
            ])
        );
        // Styling never enters Rust (author ruling, depth language).
        for word in ["shadow", "radius", "rgba", "px", "style"] {
            assert!(!js.contains(word), "the rim leaked {word} into the underlay: {js}");
        }
    }

    #[test]
    fn an_empty_layout_clears_the_auras() {
        assert_eq!(script(&[]), "window.__openlight_underlay([]);");
    }

    #[test]
    fn the_first_layout_rides_in_before_the_page_exists() {
        let js = init_script(&[Rect { x: 1.0, y: 2.0, width: 3.0, height: 4.0 }]);
        assert!(js.contains("window.__openlight_underlay ="), "the global is defined: {js}");
        assert!(js.contains(r#""width":3.0"#), "the buffer the module drains: {js}");
        // Buffering, until the module replaces it with the renderer.
        assert!(js.contains("window.__openlight_rects = rects"), "{js}");
    }
}
